### ISSUE TITLE
admin add account : request times out

### ISSUE THREAD

**Issue by damithc** [TEAMMATES/teammates#4920]
This happened 4 times in a row:

teammates.logic.core.Emails sendErrorReport: Crash report failed to send. Detailed error stack trace: 
javax.jdo.JDODataStoreException: Datastore Failure
	at org.datanucleus.jdo.NucleusJDOHelper.getJDOExceptionForNucleusException(NucleusJDOHelper.java:304)
	at org.datanucleus.jdo.JDOPersistenceManager.getObjectById(JDOPersistenceManager.java:1676)
	at org.datanucleus.jdo.JDOPersistenceManager.getObjectById(JDOPersistenceManager.java:1767)
	at org.datanucleus.store.appengine.jdo.DatastoreJDOPersistenceManager.getObjectById(DatastoreJDOPersistenceManager.java:73)
	at teammates.storage.api.AccountsDb.getAccountEntity(AccountsDb.java:212)
	at teammates.storage.api.AccountsDb.getAccount(AccountsDb.java:92)
	at teammates.storage.api.AccountsDb.getAccount(AccountsDb.java:104)
	at teammates.logic.core.AccountsLogic.isAccountAnInstructor(AccountsLogic.java:73)
	at teammates.logic.api.GateKeeper.isInstructor(GateKeeper.java:354)
	at teammates.logic.api.GateKeeper.getCurrentUser(GateKeeper.java:62)
	at teammates.logic.api.Logic.getCurrentUser(Logic.java:132)
	at teammates.ui.controller.Action.executeAndPostProcess(Action.java:343)
	at teammates.ui.controller.ControllerServlet.doPost(ControllerServlet.java:59)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:637)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:717)
	at org.mortbay.jetty.servlet.ServletHolder.handle(ServletHolder.java:511)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1166)
	at teammates.storage.datastore.DatastoreFilter.doFilter(DatastoreFilter.java:28)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.google.apphosting.utils.servlet.ParseBlobUploadFilter.doFilter(ParseBlobUploadFilter.java:125)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.google.apphosting.runtime.jetty.SaveSessionFilter.doFilter(SaveSessionFilter.java:37)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.google.apphosting.utils.servlet.JdbcMySqlConnectionCleanupFilter.doFilter(JdbcMySqlConnectionCleanupFilter.java:60)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.google.apphosting.utils.servlet.TransactionCleanupFilter.doFilter(TransactionCleanupFilter.java:43)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at org.mortbay.jetty.servlet.ServletHandler.handle(ServletHandler.java:388)
	at org.mortbay.jetty.security.SecurityHandler.handle(SecurityHandler.java:216)
	at org.mortbay.jetty.servlet.SessionHandler.handle(SessionHandler.java:182)
	at org.mortbay.jetty.handler.ContextHandler.handle(ContextHandler.java:765)
	at org.mortbay.jetty.webapp.WebAppContext.handle(WebAppContext.java:418)
	at com.google.apphosting.runtime.jetty.AppVersionHandlerMap.handle(AppVersionHandlerMap.java:260)
	at org.mortbay.jetty.handler.HandlerWrapper.handle(HandlerWrapper.java:152)
	at org.mortbay.jetty.Server.handle(Server.java:326)
	at org.mortbay.jetty.HttpConnection.handleRequest(HttpConnection.java:542)
	at org.mortbay.jetty.HttpConnection$RequestHandler.headerComplete(HttpConnection.java:923)
	at com.google.apphosting.runtime.jetty.RpcRequestParser.parseAvailable(RpcRequestParser.java:78)
	at org.mortbay.jetty.HttpConnection.handle(HttpConnection.java:404)
	at com.google.apphosting.runtime.jetty.JettyServletEngineAdapter.serviceRequest(JettyServletEngineAdapter.java:148)
	at com.google.apphosting.runtime.JavaRuntime$RequestRunnable.run(JavaRuntime.java:468)
	at com.google.tracing.TraceContext$TraceContextRunnable.runInContext(TraceContext.java:437)
	at com.google.tracing.TraceContext$TraceContextRunnable$1.run(TraceContext.java:444)
	at com.google.tracing.CurrentContext.runInContext(CurrentContext.java:256)
	at com.google.tracing.TraceContext$AbstractTraceContextCallback.runInInheritedContextNoUnref(TraceContext.java:308)
	at com.google.tracing.TraceContext$AbstractTraceContextCallback.runInInheritedContext(TraceContext.java:300)
	at com.google.tracing.TraceContext$TraceContextRunnable.run(TraceContext.java:441)
	at com.google.apphosting.runtime.ThreadGroupPool$PoolEntry.run(ThreadGroupPool.java:235)
	at java.lang.Thread.run(Thread.java:745)
NestedThrowablesStackTrace:
com.google.appengine.api.datastore.DatastoreFailureException: The thread has been interrupted and the request has timed out
	at com.google.appengine.api.datastore.FutureHelper.getInternal(FutureHelper.java:81)
	at com.google.appengine.api.datastore.FutureHelper.quietGet(FutureHelper.java:63)
	at com.google.appengine.api.datastore.DatastoreServiceImpl.get(DatastoreServiceImpl.java:41)
	at org.datanucleus.store.appengine.RuntimeExceptionWrappingDatastoreService.get(RuntimeExceptionWrappingDatastoreService.java:56)
	at org.datanucleus.store.appengine.DatastorePersistenceHandler.get(DatastorePersistenceHandler.java:100)
	at org.datanucleus.store.appengine.DatastorePersistenceHandler.get(DatastorePersistenceHandler.java:112)
	at org.datanucleus.store.appengine.DatastorePersistenceHandler.fetchObject(DatastorePersistenceHandler.java:517)
	at org.datanucleus.state.JDOStateManagerImpl.validate(JDOStateManagerImpl.java:4263)
	at org.datanucleus.ObjectManagerImpl.findObject(ObjectManagerImpl.java:2444)
	at org.datanucleus.jdo.JDOPersistenceManager.getObjectById(JDOPersistenceManager.java:1671)
	at org.datanucleus.jdo.JDOPersistenceManager.getObjectById(JDOPersistenceManager.java:1767)
	at org.datanucleus.store.appengine.jdo.DatastoreJDOPersistenceManager.getObjectById(DatastoreJDOPersistenceManager.java:73)
	at teammates.storage.api.AccountsDb.getAccountEntity(AccountsDb.java:212)
	at teammates.storage.api.AccountsDb.getAccount(AccountsDb.java:92)
	at teammates.storage.api.AccountsDb.getAccount(AccountsDb.java:104)
	at teammates.logic.core.AccountsLogic.isAccountAnInstructor(AccountsLogic.java:73)
	at teammates.logic.api.GateKeeper.isInstructor(GateKeeper.java:354)
	at teammates.logic.api.GateKeeper.getCurrentUser(GateKeeper.java:62)
	at teammates.logic.api.Logic.getCurrentUser(Logic.java:132)
	at teammates.ui.controller.Action.executeAndPostProcess(Action.java:343)
	at teammates.ui.controller.ControllerServlet.doPost(ControllerServlet.java:59)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:637)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:717)
	at org.mortbay.jetty.servlet.ServletHolder.handle(ServletHolder.java:511)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1166)
	at teammates.storage.datastore.DatastoreFilter.doFilter(DatastoreFilter.java:28)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.google.apphosting.utils.servlet.ParseBlobUploadFilter.doFilter(ParseBlobUploadFilter.java:125)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.google.apphosting.runtime.jetty.SaveSessionFilter.doFilter(SaveSessionFilter.java:37)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.google.apphosting.utils.servlet.JdbcMySqlConnectionCleanupFilter.doFilter(JdbcMySqlConnectionCleanupFilter.java:60)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.google.apphosting.utils.servlet.TransactionCleanupFilter.doFilter(TransactionCleanupFilter.java:43)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at org.mortbay.jetty.servlet.ServletHandler.handle(ServletHandler.java:388)
	at org.mortbay.jetty.security.SecurityHandler.handle(SecurityHandler.java:216)
	at org.mortbay.jetty.servlet.SessionHandler.handle(SessionHandler.java:182)
	at org.mortbay.jetty.handler.ContextHandler.handle(ContextHandler.java:765)
	at org.mortbay.jetty.webapp.WebAppContext.handle(WebAppContext.java:418)
	at com.google.apphosting.runtime.jetty.AppVersionHandlerMap.handle(AppVersionHandlerMap.java:260)
	at org.mortbay.jetty.handler.HandlerWrapper.handle(HandlerWrapper.java:152)
	at org.mortbay.jetty.Server.handle(Server.java:326)
	at org.mortbay.jetty.HttpConnection.handleRequest(HttpConnection.java:542)
	at org.mortbay.jetty.HttpConnection$RequestHandler.headerComplete(HttpConnection.java:923)
	at com.google.apphosting.runtime.jetty.RpcRequestParser.parseAvailable(RpcRequestParser.java:78)
	at org.mortbay.jetty.HttpConnection.handle(HttpConnection.java:404)
	at com.google.apphosting.runtime.jetty.JettyServletEngineAdapter.serviceRequest(JettyServletEngineAdapter.java:148)
	at com.google.apphosting.runtime.JavaRuntime$RequestRunnable.run(JavaRuntime.java:468)
	at com.google.tracing.TraceContext$TraceContextRunnable.runInContext(TraceContext.java:437)
	at com.google.tracing.TraceContext$TraceContextRunnable$1.run(TraceContext.java:444)
	at com.google.tracing.CurrentContext.runInContext(CurrentContext.java:256)
	at com.google.tracing.TraceContext$AbstractTraceContextCallback.runInInheritedContextNoUnref(TraceContext.java:308)
	at com.google.tracing.TraceContext$AbstractTraceContextCallback.runInInheritedContext(TraceContext.java:300)
	at com.google.tracing.TraceContext$TraceContextRunnable.run(TraceContext.java:441)
	at com.google.apphosting.runtime.ThreadGroupPool$PoolEntry.run(ThreadGroupPool.java:235)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.InterruptedException
	at com.google.common.util.concurrent.AbstractFuture.get(AbstractFuture.java:386)
	at com.google.common.util.concurrent.AbstractFuture$TrustedFuture.get(AbstractFuture.java:84)
	at com.google.appengine.tools.development.TimedFuture.get(TimedFuture.java:42)
	at com.google.common.util.concurrent.ForwardingFuture.get(ForwardingFuture.java:61)
	at com.google.appengine.tools.appstats.RecordingFuture.get(RecordingFuture.java:108)
	at com.google.appengine.tools.appstats.RecordingFuture$ExtendedFuture.get(RecordingFuture.java:49)
	at com.google.appengine.tools.appstats.RecordingFuture.get(RecordingFuture.java:108)
	at com.google.appengine.tools.appstats.RecordingFuture$ExtendedFuture.get(RecordingFuture.java:49)
	at com.google.appengine.api.utils.FutureWrapper.get(FutureWrapper.java:88)
	at com.google.appengine.api.datastore.AsyncDatastoreServiceImpl$5.getFutureWithOptionalTimeout(AsyncDatastoreServiceImpl.java:292)
	at com.google.appengine.api.datastore.AsyncDatastoreServiceImpl$5.aggregate(AsyncDatastoreServiceImpl.java:264)
	at com.google.appengine.api.datastore.AsyncDatastoreServiceImpl$5.get(AsyncDatastoreServiceImpl.java:235)
	at com.google.appengine.api.datastore.AsyncDatastoreServiceImpl$5.get(AsyncDatastoreServiceImpl.java:224)
	at com.google.appengine.api.utils.FutureWrapper.get(FutureWrapper.java:88)
	at com.google.appengine.api.utils.FutureWrapper.get(FutureWrapper.java:88)
	at com.google.appengine.api.utils.FutureWrapper.get(FutureWrapper.java:88)
	at com.google.appengine.api.datastore.FutureHelper.getInternal(FutureHelper.java:76)
	... 56 more

**Comment by damithc** [comment]
This is more critical than I thought at first. Seems we are getting it even on dev server. There is no way to create new accounts on the live server now.

**Comment by kanghj** [comment]
I am unable to replicate this on my dev server. In the stacktrace though, it seems that the error is occuring in GateKeeper, and not specifically on the create account action. Have there been other pages affected similarly recently?

**Comment by damithc** [comment]
yes, there are occasional 'datastore failure' severe logs, a few times a day. But this one is repeatable on the live server. I simply can't create new accounts now. Noticed it a few times during dev server testing (in admin add account test).

**Comment by damithc** [comment]
Emailed you the full log. Too much logging could be the reason?

**Comment by LowWeiLin** [comment]
Unable to reproduce on dev, but it does take a really long time. Maybe we could disable creating the demo course as a quick fix?

**Comment by LowWeiLin** [comment]
I think i found the issue. This loop in BackDoorLogic persistDataBundle() was doing unnecessary work,
which made creating the demo course 3-4 times slower. This was introduced in PR #4818...
```
for (FeedbackResponseAttributes response : responses.values()) {
    updateRespondants(response.feedbackSessionName, response.courseId);
}
```

**Comment by bamboo3250** [comment]
I also think this is timeout issue. I set a timer and it got interrupted after 60s. 
`NestedThrowablesStackTrace:
com.google.appengine.api.datastore.DatastoreFailureException: The thread has been interrupted and the request has timed out
`

**Comment by kanghj** [comment]
@bamboo3250, do you want to fix this? Fix this or #4904, the lines, mentioned by @LowWeiLin and in the other issue, causing this timeout can be optimised, extract out all the session names and course ids from the responses before doing the backdoor operation

**Comment by bamboo3250** [comment]
@kanghj I don't think I'm ready to do a hot fix on Feedback Response area atm. 
I just want to mention that we had the same issue a few weeks ago #4784

**Comment by kanghj** [comment]
@bamboo3250, ok I did remember something like that but couldn't find it. Perhaps it's caused by too many writes in a short time period time.  

@unyoungwax, do you want to fix this?

**Comment by wkurniawan07** [comment]
Joining the party quite late.
The update respondents loop has been there since long ago, but changing from iterating sessions to iterating responses (which indeed was done #4818) did make it slower. I don't think we can just claim it "unnecessary".

**Comment by damithc** [comment]
@unyoungwax we need an immediate fix. If you are not sure you can reach a fix within 24 hours, please let us know so that someone else can take the issue.

**Comment by damithc** [comment]
Just realized that I can use an older version on the live server (it has the last 10 versions) to overcome this issue. So it is no longer critical. But fix it soon, at least by V5.72
